File: frontend/src/types/rule.ts
import { Script, Snapshot } from "@/types/snapshot";
import { YaraMatch } from "@/types/yara";

export type TargetTypes = "html" | "whois" | "certificate" | "script";

interface RuleOptions {
  allowedNetworkAddresses: string | undefined;
  disallowedNetworkAddresses: string | undefined;
  allowedResourceHashes: string | undefined;
  disallowedResourceHashes: string | undefined;
}

export interface Rule extends RuleOptions {
  id: string;
  name: string;
  target: TargetTypes;
  source: string;
  snapshots: Snapshot[];
  createdAt: string;
  updatedAt: string;
}

export interface RuleFilters {
  name: string | undefined;
  target: TargetTypes | undefined;
  source: string | undefined;
}

export interface Match {
  id: string;
  snapshot: Snapshot;
  rule: Rule;
  script: Script | null;
  matches: YaraMatch[];
  createdAt: string;
}

export interface MatchFilters {
  snapshotId: string | undefined;
  ruleId: string | undefined;
  fromAt: Date | undefined;
  toAt: Date | undefined;
}

export interface CreateRulePayload extends RuleOptions {
  name: string;
  target: string;
  source: string;
}

export interface UpdateRulePayload extends RuleOptions {
  name: string | undefined;
  target: string | undefined;
  source: string | undefined;
}

File: app/factories/rule.py
from typing import TYPE_CHECKING

from app import schemas

if TYPE_CHECKING:
    from app.models.rule import Rule


def build_snapshots(rule: "Rule"):
    if hasattr(rule, "related_snapshots"):
        snapshots = rule.related_snapshots or []
        return [snapshot.to_model() for snapshot in snapshots]

    return []


class RuleFactory:
    @staticmethod
    def from_model(rule: "Rule") -> schemas.Rule:
        return schemas.Rule(
            id=rule.id,
            name=rule.name,
            source=rule.source,
            target=rule.target,
            allowed_network_addresses=rule.allowed_network_addresses,
            disallowed_network_addresses=rule.disallowed_network_addresses,
            allowed_resource_hashes=rule.allowed_resource_hashes,
            disallowed_resource_hashes=rule.disallowed_resource_hashes,
            created_at=rule.created_at,
            updated_at=rule.updated_at,
            snapshots=build_snapshots(rule),
        )
